tfworker/util/copier.py

```diff
@@ -0,0 +1,217 @@
+
+from tfworker.util.system import pipe_exec
+
+
+class CopyFactory:
+    """ The factory class for creating copiers """
+
+    registry = {}
+
+    @classmethod
+    def register(cls, name: str) -> Callable:
+        """ Class method to register copiers """
+
+        def inner_wrapper(wrapped_class: Copier) -> Callable:
+            if name in cls.registry:
+                raise ValueError(f"Executor {name} already exists")
+            cls.registry[name] = wrapped_class
+            return wrapped_class
+
+        return inner_wrapper
+
+    @classmethod
+    def create(cls, source: str, **kwargs) -> "Copier":
+        """ create returns a copier type that supports handling the provided source """
+        copier_class = cls.registry[cls.get_copier_type(source, **kwargs)]
+        copier = copier_class(source, **kwargs)
+        return copier
+
+    @classmethod
+    def get_copier_type(cls, source: str, **kwargs) -> str:
+        """ get_copier_type tries to find a supported copier based on the provided source """
+        for copier_type, copier_class in cls.registry.items():
+            if copier_class.type_match(source, **kwargs):
+                return copier_type
+        raise NotImplementedError(f"no valid copier for {source}")
+
+
+class Copier(metaclass=ABCMeta):
+    """ The base class for definition copiers """
+
+    def __init__(self, source: str, **kwargs):
+        self._source = source
+
+        for k, v in kwargs.items():
+            setattr(self, f"_{k}", v)
+
+        if hasattr(self, "_conflicts"):
+            if type(self._conflicts) is not list:
+                raise ValueError("Conflicts must be a list of filenames to disallow")
+
+    @abstractstaticmethod
+    def type_match(source: str, **kwargs) -> bool:  # pragma: no cover
+        """ type_match determins if the source is supported/handled by a copier """
+        pass
+
+    @abstractmethod
+    def copy(self, **kwargs) -> bool:  # pragma: no cover
+        """ copy executes the copy from the source, into the working path """
+        pass
+
+    @property
+    def root_path(self):
+        """ root_path returns an optional root path to use for relative file operations """
+        if hasattr(self, "_root_path"):
+            return self._root_path
+        else:
+            return ""
+
+    @property
+    def conflicts(self):
+        """ conflicts returns a list of disallowed files """
+        if hasattr(self, "_conflicts"):
+            return self._conflicts
+        else:
+            return []
+
+    @property
+    def source(self):
+        """ source contains the source path providede """
+        return self._source
+
+    def get_destination(self, make_dir: bool = True, **kwargs) -> str:
+        """ get_destination returns the destination path, and optionally makes the destinatination directory """
+        if not (hasattr(self, "_destination") or "destination" in kwargs.keys()):
+            raise ValueError("no destination provided")
```

Comment:
This is basically checking that either the subclass has a `_destination` attribute or `destination` was passed through kwargs? Is it a requirement that subclasses of `Copier` have a `_destination` attribute? I wonder if that's something you can enforce with python's type system. I know some languages support that but not sure if python's type system does. Not really important, just thinking out loud that it would be nice if there was a way to check this with mypy.

Reply:
I left destination as optional during the constructor so one could use the same copier to copy to multiple destinations. It can be set at either Copier instantiation or during the copy() method, that's why I put the tests around it since if you try a copy with it never set it will raise an error, but I wanted to leave it flexible in case it gets reused in other parts of the code. Could probably simplify and keep a more consistent interface though.